src/core/server/server.cc

```diff
@@ -910,7 +1133,9 @@ void Server::AddListener(OrphanablePtr<ListenerInterface> listener) {
     channelz_node_->AddChildListenSocket(
         listen_socket_node->RefAsSubclass<channelz::ListenSocketNode>());
   }
-  listeners_.emplace_back(std::move(listener));
+  ListenerInterface* ptr = listener.get();
+  listeners_.emplace_back(this, std::move(listener));
+  ptr->SetServerListenerWrapper(&listeners_.back());
```

Comment:
> Also, if we are always setting this on the ListenerInterface impl, can we somehow pass it in to the ctor instead of needing a method to set it after construction?

We create the actual listener implementation, for example, when we add a port to the server. After a listener is created, we register it with the server which is when we create a `ListenerState`. 

One other way I can think of is to pre-create the `ListenerState` with another method on the server, and once the listener is created, we register the listener with the `ListenerState`.